src/riscv/lib/src/pvm/tezos.rs

```diff
@@ -359,19 +413,26 @@ pub(super) fn handle_tezos<MC, M>(
     MC: MemoryConfig,
     M: ManagerReadWrite,
 {
-    // TODO: RV-777: remove below and instead have each system call return a `ProgramCounterUpdate`
-    let pc = machine.hart.pc.read().wrapping_add(4);
-    machine.hart.pc.write(pc);
-
     let sbi_function = machine.hart.xregisters.read(a6);
-    match sbi_function {
+    let TezosCallResult { status, pc_update } = match sbi_function {
         SBI_TEZOS_INBOX_NEXT => handle_tezos_inbox_next(status),
-        SBI_TEZOS_ED25519_SIGN => sbi_wrap(machine, handle_tezos_ed25519_sign),
-        SBI_TEZOS_ED25519_VERIFY => sbi_wrap(machine, handle_tezos_ed25519_verify),
-        SBI_TEZOS_BLAKE2B_HASH256 => sbi_wrap(machine, handle_tezos_blake2b_hash256),
-        SBI_TEZOS_SECP256K1_VERIFY => sbi_wrap(machine, handle_tezos_secp256k1_verify),
-        SBI_TEZOS_KECCAK256_HASH => sbi_wrap(machine, handle_tezos_keccak256_hash),
+        SBI_TEZOS_ED25519_SIGN => TezosCallResult::incr_pc(handle_tezos_ed25519_sign(machine)),
+        SBI_TEZOS_ED25519_VERIFY => TezosCallResult::incr_pc(handle_tezos_ed25519_verify(machine)),
+        SBI_TEZOS_BLAKE2B_HASH256 => {
+            TezosCallResult::incr_pc(handle_tezos_blake2b_hash256(machine))
+        }
+        SBI_TEZOS_SECP256K1_VERIFY => {
+            TezosCallResult::incr_pc(handle_tezos_secp256k1_verify(machine))
+        }
+        SBI_TEZOS_KECCAK256_HASH => TezosCallResult::incr_pc(handle_tezos_keccak256_hash(machine)),
         SBI_TEZOS_REVEAL => handle_tezos_reveal(machine, reveal_request, status),
-        _ => handle_not_supported(&mut machine.hart.xregisters),
+        _ => handle_not_supported(),
+    };
+
+    machine.update_pc(machine.hart.pc.read(), pc_update);
+    match status {
+        CallStatus::Success(reg_val) => sbi_return1(&mut machine.hart.xregisters, reg_val),
+        CallStatus::Error(error) => sbi_return_error(&mut machine.hart.xregisters, error),
+        CallStatus::Suspended => {}
     }
```

Comment:
This part is more or less the same in what it does to the general `handle_system_call` function in `linux.rs`. Do you think you can unify these behaviours? What I see as the source for this maybe unnecessary duplication is the fact that for tezos functions we return early in the match statement [here](https://github.com/tezos/riscv-pvm/blob/59fdecc5982111e7051e8400ab18ea43db2499e6/src/riscv/lib/src/pvm/linux.rs#L782)